speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backend/Mediator.inl

```diff
@@ -24,27 +24,52 @@ void Mediator<DecisionMaker>::SetInterventionType(INTERVENTION_TYPE p_type)
 }
 
 template<typename DecisionMaker>
-void Mediator<DecisionMaker>::DriveTick(tCarElt* p_car, tSituation* p_situation) {}
+void Mediator<DecisionMaker>::DriveTick(tCarElt* p_car, tSituation* p_situation)
+{
+    DriveSituation currentSituation(
+        m_environment,
+        CarInfo(
+            TrackPosition(false, p_car->pub.trkPos.toStart, p_car->pub.trkPos.toRight, p_car->pub.trkPos.toMiddle, p_car->pub.trkPos.toLeft),
+            p_car->pub.speed, p_car->race.topSpeed, p_car->priv.gear, false),
+        PlayerInfo(p_car->ctrl.steer, p_car->ctrl.accelCmd, p_car->ctrl.brakeCmd, p_car->ctrl.clutchCmd));
+
+    m_decisionMaker.Decide(currentSituation);
+}
 
 template<typename DecisionMaker>
-void Mediator<DecisionMaker>::RaceStart(tTrack* p_track, void* p_carHandle, void** p_carParmHandle, tSituation* p_situation) {}
+void Mediator<DecisionMaker>::RaceStart(tTrack* p_track, void* p_carHandle, void** p_carParmHandle, tSituation* p_situation)
+{
+    m_environment.SetTimeOfDay(p_track->local.timeofday);
+    m_environment.SetClouds(p_track->local.clouds);
+    m_environment.SetRain(p_track->local.rain);
+}
 
 template<typename DecisionMaker>
 void Mediator<DecisionMaker>::RaceStop() {}
 
 template<typename DecisionMaker>
 DriveSituation* Mediator<DecisionMaker>::Simulate()
 {
+    
     return nullptr; 
 }
 
 template<typename DecisionMaker>
-Mediator<DecisionMaker>& Mediator<DecisionMaker>::GetInstance() {
-    static Mediator<DecisionMaker> s_instance;
-    return s_instance;
+Mediator<DecisionMaker>* Mediator<DecisionMaker>::GetInstance() {
+    if(m_instance == nullptr)
+    {
+        std::string pointerName("00000000");
+        std::ifstream file("Singletons/Mediator");
+        getline(file, pointerName);
```

Comment:
Would it be possible to check whether there is already a pointer in the file, and if not, this class adds it?
Then the next module that tries to get the singleton instance **does** find a pointer in the file and just returns that.
I think this would render the PointerDistributor class obsolete, which reduces the complexity of making a singleton.

Below some pseudocode:
```suggestion
        std::fstream file("Singletons/Mediator");
        if(file is empty)
        {
           create instance
           write instance pointer to file
           close file
           return instance 
        }
        else // file non empty -> contains pointer
        { 
           read pointer from file
           close file
           cast into meditor instance
           return instance
        }
```

Reply:
in theory this might be possible with a destructor and I will try that later today

Reply:
as the calling of a destructor is not guaranteed, it would only be possibe by wiping the singletons folder. I will implement this today.